File: code/WMS_Model.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Feb 19 13:02:54 2019
@author: saisu
"""
from scipy.io import loadmat
import math as m
import numpy as np
import matplotlib.pyplot as plt
import nf1f_WMSfit
import scipy.signal as signal
from scipy.signal import lfilter
import scipy.optimize as optimization
import time
import warnings

warnings.filterwarnings("ignore")

def downsampling(data,space):
    new_list=[]
    k=int(len(data)/space)
    for i in range(k):
        new_list.append(data[i*int(space)])
    return new_list
    

################################# Main #############################
num=0
H2O_database= loadmat("/Users/saisu/Desktop/Final Year Projrct/H2O_300K.mat")
M=18 #molecule H2O
P=1.00 #pressure 1 atm
T=300 #temperature K (must fixed here)
x_s=0.02 #mole fraction of H2O
x_air=1-x_s # mole fraction of air
L=50 #pathlength cm
v0=7407.8  #center frequency cm-1
res_v=0.0002  #spectra resolution in calculation;
v0_L=1.00
v0_R=1-res_v    #selected spectra range in calculation; v0_L: range left to v0;  v0_R: range right to v0
value_list=list(H2O_database.values())
H2O_300K=value_list[3] 
col1_v=H2O_300K[:,0].tolist(); col2_St=H2O_300K[:,1].tolist();gamma_air=H2O_300K[:,2].tolist();gamma_self=H2O_300K[:,3].tolist();col5_E=H2O_300K[:,4].tolist();

index_v=list()  # initiate empty list
index_St=list()
index_gamma_air=list()
index_gamma_self=list()
index_E=list()
wD=list()
wC=list()
wV=list()
drt_ved=list()
beta=list()
faiv0=list()
#for i in H2O_300K[:,0]: #iterate all v, i.e col1:v
for i in range(len(col1_v)):
   if col1_v[i]>=v0-v0_L and col1_v[i]<=v0+v0_R:
     index_v.append(col1_v[i])  # find the transitions in this range
     index_St.append(col2_St[i]) 
     index_gamma_air.append(gamma_air[i]) 
     index_gamma_self.append(gamma_self[i])
     index_E.append(col5_E[i]) 
     
##############     calculate the parameters of the selected transitions     ###############
#for i in range(len(col1_v)):
for i in range(len(index_v)):
     wD.append(np.multiply(7.1623e-7,index_v[i]*(T/M)**0.5))        #Dopple width
     wC.append(P*2*(x_s*index_gamma_self[i]+x_air*index_gamma_air[i]))      #Collision width
     wV.append(0.5346*wC[i]+(0.2166*wC[i]**2+wD[i]**2)**0.5)                 #Voigt width
     drt_ved.append(wD[i]/2/m.sqrt(m.log(2)))
     beta.append(drt_ved[i]/(wC[i]/2+drt_ved[i]))
     faiv0.append((beta[i]/drt_ved[i]/m.sqrt(m.pi))+2*((1-beta[i])/(m.pi*wC[i])))
 
# v before modulation
v=np.linspace(v0-v0_L,v0+v0_R,10000)
# v after modulation
t=np.linspace(0,1,len(v))
fm=350          #modulation  frequency
N=len(v)/fm     #number of samples per cycle
fai_0=0.5*m.pi  #phase shift of FM
a=0.1           #modulation depth
temp=2*m.pi*fm*t+fai_0
temp=[m.cos(x) for x in temp]
v_mod=v+np.dot(a,temp) 

'''
fig = plt.figure()
ax = fig.add_subplot(111)
ax.plot(t,v_mod)
plt.xlabel('Time/s')
plt.ylabel('Frequency/cm-1')
plt.title('Frequency Response to the Laser Injection Current')
plt.show()
'''

###################    DAS voigt lineshape     ##########################
vv_DAS_per_v=np.mat(np.zeros((len(index_v),len(v))))

for j in range(len(index_v)): #20  
    temp1=-4*m.log(2)*((v-index_v[j])/wV[j])**2
    temp1=[m.exp(x) for x in temp1]
    temp2=-0.4*abs(abs((v-index_v[j])/wV[j])**2.25)
    temp2=[m.exp(x) for x in temp2]
    vv_DAS_per_v[j,:]=(P*x_s*L*index_St[j]*7.34e21/T*(np.multiply((1-wC[j]/wV[j]),temp1)+wC[j]/wV[j]/(4*((v-index_v[j])/wV[j])**2+1)+0.016*(1-wC[j]/wV[j])*wC[j]/wV[j]*(temp2-10/(abs(abs((v-index_v[j])/wV[j])**2.25)+10)))*faiv0[j])
    damn=np.trapz(vv_DAS_per_v[j,:],v)
    vv_DAS_per_v[j,:]=np.multiply(vv_DAS_per_v[j,:],P*x_s*L*index_St[j]*7.34e21/T/np.trapz(vv_DAS_per_v[j,:],v))
absorb_DAS=np.array( vv_DAS_per_v.sum(axis=0))

        
#Figure Plot    
'''
plt.figure()
plt.plot(v,absorb_DAS,label="Simulation (Voigt)")
plt.legend(loc='upper left')
plt.text(7406.85, 0.014, "Molecule: H2O", size = 12, alpha = 0.9)
plt.text(7406.85, 0.012, "T: 300K", size = 12, alpha = 0.9)
plt.text(7406.85, 0.010, "L: 50cm", size = 12, alpha = 0.9)
plt.text(7406.85, 0.008, "P: 1atm", size = 12, alpha = 0.9)
plt.text(7406.85, 0.006, "v0: 7407.8cm-1", size = 12, alpha = 0.9)
plt.text(7406.85, 0.004, "X: 2%", size = 12, alpha = 0.9)
plt.xlabel('Frequency/cm-1')
plt.ylabel('Absorbance')
plt.title('Absorbance vs. Frequency')
plt.show()
'''

#################       Modulated DAS voigt lineshape      #############################

vv_DAS_modulated=np.mat(np.zeros((len(index_v),len(v_mod))))

for j in range(len(index_v)): #20  
    temp1=-4*m.log(2)*((v_mod-index_v[j])/wV[j])**2
    temp1=[m.exp(x) for x in temp1]
    temp2=-0.4*abs(abs((v_mod-index_v[j])/wV[j])**2.25)
    temp2=[m.exp(x) for x in temp2]
    vv_DAS_modulated[j,:]=(P*x_s*L*index_St[j]*7.34e21/T*(np.multiply((1-wC[j]/wV[j]),temp1)+wC[j]/wV[j]/(4*((v_mod-index_v[j])/wV[j])**2+1)+0.016*(1-wC[j]/wV[j])*wC[j]/wV[j]*(temp2-10/(abs(abs((v_mod-index_v[j])/wV[j])**2.25)+10)))*faiv0[j])
    vv_DAS_modulated[j,:]=np.multiply(vv_DAS_modulated[j,:],P*x_s*L*index_St[j]*7.34e21/T/np.trapz(vv_DAS_modulated[j,:],v_mod))
absorb_modulated=np.array(vv_DAS_modulated.sum(axis=0)).T


#Figure Plot    
'''
plt.figure()
plt.plot(t,absorb_modulated,label="Simulation")
plt.legend(loc='upper left')
plt.xlabel('Time/s')
plt.ylabel('Absorbance')
plt.title('Modulated Absorbance vs. Time')
plt.show()
'''
        
#######################  I0 after modulation   #########################
I_max=0.9
I_min=0
k=(I_max-I_min)/1.2
I0_aver=I_min+np.dot(k,(t+0.2))  #ramp base
i0=0.1              #amplitude of linear IM
fai_1=-0.2*m.pi          #phase shift of FM/IM
i2=0                #amplitude of nonlinear IM
fai_2=1.31*m.pi     #phase shift of nonlinear IM
temp=2*m.pi*fm*t+fai_1
temp=[m.cos(x) for x in temp]
temp2=4*m.pi*fm*t+fai_2
temp2=[m.cos(x) for x in temp2]
I0=I0_aver*(1+i0/I0_aver*temp+i2/I0_aver*temp2)
absorb_modulated=np.multiply(-1,absorb_modulated)
tempt=[m.exp(x) for x in absorb_modulated]
It=np.multiply(I0,tempt) #transimitted laser intensity
 #Figure Plot    

plt.figure()
plt.plot(t,It,label="Simulation")  
plt.title('Absorption-free Laser Intensity vs. time')     


###############   Lock-in detection    #################################
sig=It
n=np.linspace(1,len(t),len(t))

# 2f
harmonic=2
temp=harmonic*2*m.pi*n/N
Vrs=[m.sin(x) for x in temp]
Vrc=[m.cos(x) for x in temp]
x2=np.multiply(sig,Vrs)
y2=np.multiply(sig,Vrc)
#b, a = signal.butter(4,1/7.6/N, 'low')
order, wn = signal.buttord(1/8/N, 1/2/N,gpass=1/3.5,gstop=60, analog='true')
b, a = signal.butter(order,wn, 'low')
X2 = lfilter(b,a, x2)
Y2 = lfilter(b,a, y2)
tempt=X2**2+Y2**2
tempt=[m.sqrt(x) for x in tempt]
S2=np.dot(2,tempt)

# 1f
harmonic=1
temp=harmonic*2*m.pi*n/N
Vrs=[m.sin(x) for x in temp]
Vrc=[m.cos(x) for x in temp]
x1=np.multiply(sig,Vrs)
y1=np.multiply(sig,Vrc)
X1 = lfilter(b,a, x1)
Y1 = lfilter(b,a, y1)
tempt=X1**2+Y1**2
tempt=[m.sqrt(x) for x in tempt]
S1=np.dot(2,tempt)
norm_2f=[S2[k]/S1[k] for k in range(len(S2))]

'''
plt.figure()
plt.plot(v[int(len(t)/5): int(len(t))],norm_2f[int(len(t)/5): int(len(t))],label="Simulation")  
plt.legend(loc='upper left')
plt.xlabel('Frequency/ cm-1')
plt.ylabel('S2f/1f')
plt.title('1f-noralized 2f signal  vs. frequency')     
'''

###############    fitting the CF-WMS 2f/1f    ###############################

x0=[0.05, 0.1, 0.001, 0, 0.05,v0]          #Starting guess
lb=[0.01, 0.08, 0.0001, -1*m.pi, 0.05, v0-v0_L]    #lower bound
ub=[0.1, 0.15, 0.1, m.pi, 0.2, v0+v0_R]     #upper bound
    #   x is the vector of the fitting parameters, x(0)=wD; x(1)=wC;
    #   x(2)=P*X*S(T)*L; x(3)=fai_1; x(4)=a; x(5)=v0
bdy=[lb,ub] # boundary

input_data=v
output_data=norm_2f[int(len(t)/5): int(len(t))-int(len(t)/5)]

new_in=downsampling(input_data, 4)
new_out=downsampling(output_data,4)

start_3 = time.clock()
popt,pconv=optimization.curve_fit(nf1f_WMSfit.nf1f_WMSfit,new_in,new_out,x0,bounds=bdy,ftol=1e-1)
x=popt
elapsed_3 = (time.clock() - start_3)
print("Time used for Fitting:",elapsed_3)

###############    fitting the norm_2f_1f with fitted parameters    ###############################
norm_2f_fitted=nf1f_WMSfit.nf1f_WMSfit(v,popt[0],popt[1],popt[2],popt[3],popt[4],popt[5])

###############    fitting the absorbance with fitted parameters    ###############################

wV=0.5346*x[1]+(0.2166*x[1]**2+x[0]**2)**0.5
drt_ved=x[0]/2/(m.sqrt(m.log(2)))
beta=drt_ved/(x[1]/2+drt_ved)  
faiv=(beta/drt_ved/m.sqrt(m.pi))+2*((1-beta)/(m.pi*x[1]))
t=np.linspace(0,1,len(v))
fm=350            #modulation frequency
fai_0=0.5*m.pi     #phase shift of FM
a=x[4]              #modulation depth
temp=2*m.pi*fm*t+fai_0
temp=[m.cos(x) for x in temp]
v_mod=v+np.dot(a,temp) 

tempt=(np.multiply(-4*m.log(2),((v_mod-x[5])/wV)**2))
tempt2=-0.4*abs(abs((v_mod-x[5])/wV)**2.25)
tempt=[m.exp(x) for x in tempt]
tempt2=[m.exp(x) for x in tempt2]
vv_mod=(np.dot((1-x[1]/wV),tempt)+\
x[1]/wV/(4*((v_mod-x[5])/wV)**2+1)+\
0.016*(1-x[1]/wV)*x[1]/wV *(tempt2-10/(abs(abs(v_mod-x[5]/wV)**2.25)+10)))*faiv
absorb_modulated=x[2]*vv_mod    #modulated absorbance

tempt=(np.multiply(-4*m.log(2),((v-x[5])/wV)**2))
tempt2=-0.4*abs(abs((v-x[5])/wV)**2.25)
tempt=[m.exp(x) for x in tempt]
tempt2=[m.exp(x) for x in tempt2]
vv=(np.dot((1-x[1]/wV),tempt)+\
x[1]/wV/(4*((v-x[5])/wV)**2+1)+\
0.016*(1-x[1]/wV)*x[1]/wV *(tempt2-10/(abs(abs(v-x[5]/wV)**2.25)+10)))*faiv
absorb_fit=x[2]*vv              #fitted absorbance



#Figure Plot 
plt.figure()
plt.plot(norm_2f[int(len(t)/5): int(len(t))-int(len(t)/5)],color='r',label="Measured")
plt.plot(norm_2f_fitted,color='b',label="Fitted")
plt.legend(loc='upper left')
plt.xlabel('Frequency/ cm-1')
plt.ylabel('S2f/1f')
plt.title('Fitted S2f/1f vs. Measured S2f/1f')  

    
    

    
    
    
    
    
    
    
    
    

File: code/WMS_Model.m
% WMS for selected H2O spectra 
close all;
clear all; clc;
load H2O_300K.mat; %col1:v; col2:S(300k); col3:gamma_air; col4:gamma_self; col5:E"
M=18; %molecule H2O
P=1; %pressure 1 atm
T=300; %temperature K (must fixed here)
x_s=0.02; % mole fraction of H2O
x_air=1-x_s; % mole fraction of air
L=50; % pathlength cm
global num;
num=0;
global damn;
damn=0;
%% plot modulated v(t) 
v0=7407.8; %center frequency cm-1
res_v=0.0002; %spectra resolution in calculation;
v0_L=1; v0_R=1-res_v; %selected spectra range in calculation; v0_L: range left to v0;  v0_R: range right to v0
index_v=find((H2O_300K(:,1)>=v0-v0_L)&(H2O_300K(:,1)<=v0+v0_R)); % find the transitions in this range

%calculate the parameters of the selected transitions
wD=(7.1623e-7)*H2O_300K(index_v,1)*(T/M).^0.5; % Dopple width
wC=P*2*(x_s*H2O_300K(index_v,4)+x_air*H2O_300K(index_v,3)); %Collision width
wV=0.5346*wC+(0.2166*wC.^2+wD.^2).^0.5; %Voigt width
drt_ved=wD/2/sqrt(log(2));
beta=drt_ved./(wC/2+drt_ved);
faiv0=(beta./drt_ved/sqrt(pi))+2*((1-beta)./(pi*wC));   % Peak height of the Voigt Lineshape

%v before modulation
v=v0-v0_L:res_v:v0+v0_R;

%v after modulation
t=linspace(0,1,length(v));
fm=200 %modulation  frequency
N=length(v)/fm %number of samples per cycle
fai_0=0.5*pi;  %phase shift of FM
a=0.1; %modulation depth
v_mod=v+a*cos(2*pi*fm*t+fai_0);
% plot(t,v_mod)

% DAS voigt lineshape
for i=1:length(index_v)
    vv_DAS(:,i)=P*x_s*L*H2O_300K(index_v(i),2)*7.34e21/T*((1-wC(i)./wV(i)).*exp(-4*log(2)*((v-H2O_300K(index_v(i),1))./wV(i)).^2)+wC(i)./wV(i)./(4*((v-H2O_300K(index_v(i),1))./wV(i)).^2+1)+0.016*(1-wC(i)./wV(i))*wC(i)./wV(i).*(exp(-0.4*abs(((v-H2O_300K(index_v(i),1))./wV(i)).^2.25))-10./(abs(((v-H2O_300K(index_v(i),1))./wV(i)).^2.25)+10)))*faiv0(i);
    vv_DAS(:,i)=vv_DAS(:,i)*P*x_s*L*H2O_300K(index_v(i),2)*7.34e21/T/trapz(v,vv_DAS(:,i));
end
absorb_DAS=sum(vv_DAS,2);
 %plot(v,absorb_DAS); hold on;
% xlabel('Wavenumber [cm-1]');ylabel('Absorbance')
A_DAS=trapz(v,absorb_DAS);

% modulated voigt lineshape
for i=1:length(index_v)
    vv(:,i)=P*x_s*L*H2O_300K(index_v(i),2)*7.34e21/T*((1-wC(i)./wV(i)).*exp(-4*log(2)*((v_mod-H2O_300K(index_v(i),1))./wV(i)).^2)+wC(i)./wV(i)./(4*((v_mod-H2O_300K(index_v(i),1))./wV(i)).^2+1)+0.016*(1-wC(i)./wV(i))*wC(i)./wV(i).*(exp(-0.4*abs(abs((v_mod-H2O_300K(index_v(i),1))./wV(i)).^2.25))-10./(abs(abs((v_mod-H2O_300K(index_v(i),1))./wV(i)).^2.25)+10)))*faiv0(i);
    vv(:,i)=vv(:,i)*P*x_s*L*H2O_300K(index_v(i),2)*7.34e21/T/trapz(v,vv(:,i));
end
absorb=sum(vv,2);
% plot(t,absorb); 

%I0 after modulation
I_max=0.9;
I_min=0;
k=(I_max-I_min)/1.2;
I0_aver=I_min+k.*(t+0.2); %ramp base
% I0_aver=0.4*cos(2*pi*1/2*t)+0.5; %sine base
i0=0.1; %amplitude of linear IM
fai_1=-0.2*pi; %phase shift of FM/IM
i2=0;%amplitude of nonlinear IM
fai_2=1.31*pi;%phase shift of nonlinear IM
I0=I0_aver.*(1+i0./I0_aver.*cos(2*pi*fm*t+fai_1)+i2./I0_aver.*cos(4*pi*fm*t+fai_2));
% figure(1); plot(t,I0); hold on;

It=I0.*exp(-absorb'); %transimitted laser intensity
% It=awgn(It,60,'measured'); %add Guassian white noise to the transmission
figure(2); plot(t,It);
 xlabel('Time');ylabel('Transmission')

%% the following program are the lock-in
sig=It;
xx=1:length(t);
harmonic=2;
Vrs=sin(harmonic*2*pi.*xx/N);
Vrc=cos(harmonic*2*pi.*xx/N);
x2=sig.*Vrs;
y2=sig.*Vrc;
d=fdesign.lowpass('Fp,Fst,Ap,Ast',1/4/N,1/N,1,60);
Hd = design(d,'butter');
X2= filter(Hd,x2);
Y2=filter(Hd,y2);
S2=2*sqrt(X2.^2+Y2.^2);
%figure();plot(t,S2);hold on;

harmonic=1;
Vrs=sin(harmonic*2*pi.*xx/N);
Vrc=cos(harmonic*2*pi.*xx/N);
x1=sig.*Vrs;
y1=sig.*Vrc;
X1= filter(Hd,x1);
Y1=filter(Hd,y1);
S1=2*sqrt(X1.^2+Y1.^2);
% plot(t,S1);

% freq=0:(2*pi)/length(x1):pi;
% x1dft=fft(x2);
% X1dft=fft(X2);
% figure
% plot(freq/pi,abs(x1dft(1:length(x1)/2+1)))
% hold on
% plot(freq/pi,abs(X1dft(1:length(X1)/2+1)))

norm_2f=S2./S1;
%figure()
%plot(t,norm_2f);
%5xlabel('Time');ylabel('S 2f/1f')

%% fitting the CF-WMS 2f/1f
x0=[0.05; 0.1;0.001;0;0.05; v0]; % Starting guess
lb=[0.01; 0.08;1e-4;-pi;0.05; v0-v0_L]; %lower bound
ub=[0.1; 0.15;0.1;pi;0.2; v0+v0_R]; %upper bound
%x is the vector of the fitting parameters, x(1)=wD; x(2)=wC;
%x(3)=P*X*S(T)*L; x(4)=fai_1; x(5)=a; x(6)=v0
options = optimoptions(@lsqcurvefit,'MaxIter',1500);
tic;
[x,resnorm,residual] = lsqcurvefit(@nf1f_WMSfit,x0,v,norm_2f((length(t)/5):end),lb,ub,options); 
toc;
wV_fit=0.5346*x(2)+(0.2166*x(2)^2+x(1)^2)^0.5;
drt_ved_fit=x(1)/2/sqrt(log(2));
beta_fit=drt_ved_fit./(x(2)/2+drt_ved_fit);
faiv_fit=(beta_fit./drt_ved_fit/sqrt(pi))+2*((1-beta_fit)./(pi*x(2)));
v_mod_fit=v+x(5)*cos(2*pi*fm*t+fai_0);
vv_fit=((1-x(2)/wV_fit)*exp(-4*log(2)*((v_mod_fit-x(6))/wV_fit).^2)+x(2)./wV_fit./(4*((v_mod_fit-x(6))./wV_fit).^2+1)+0.016*(1-x(2)/wV_fit)*x(2)/wV_fit.*(exp(-0.4*abs(((v_mod_fit-x(6))/wV_fit).^2.25))-10./(abs(((v_mod_fit-x(6))/wV_fit).^2.25)+10)))*faiv_fit;
absorb_fit=x(3)*vv_fit;
a_fit=x(3)*((1-x(2)/wV_fit)*exp(-4*log(2)*((v-x(6))/wV_fit).^2)+x(2)./wV_fit./(4*((v-x(6))./wV_fit).^2+1)+0.016*(1-x(2)/wV_fit)*x(2)/wV_fit.*(exp(-0.4*abs(((v-x(6))/wV_fit).^2.25))-10./(abs(((v-x(6))/wV_fit).^2.25)+10)))*faiv_fit;%calculate fitted lineshape
A_fit=trapz(v,a_fit);%calculate fitted integrated absorbance

figure();
plot(v,absorb_DAS,'color','r'); hold on;
xlabel('Wavenumber [cm-1]');ylabel('Absorbance')
plot(v,a_fit,'color','b');
xlabel('Wavenumber [cm-1]');ylabel('Absorbance')

I0_fit=I0_aver.*(1+i0./I0_aver.*cos(2*pi*fm*t+x(4))+i2./I0_aver.*cos(4*pi*fm*t+fai_2));
It_fit=I0_fit.*exp(-absorb_fit);
sig_fit=It_fit;

harmonic=2;
Vrs=sin(harmonic*2*pi.*xx/N);
Vrc=cos(harmonic*2*pi.*xx/N);
x2_fit=sig_fit.*Vrs;
y2_fit=sig_fit.*Vrc;
d_fit=fdesign.lowpass('Fp,Fst,Ap,Ast',1/4/N,1/N,1,60);
Hd_fit = design(d_fit,'butter');
X2_fit= filter(Hd_fit,x2_fit);
Y2_fit=filter(Hd_fit,y2_fit);
S2_fit=2*sqrt(X2_fit.^2+Y2_fit.^2);

harmonic=1;
Vrs_fit=sin(harmonic*2*pi.*xx/N);
Vrc_fit=cos(harmonic*2*pi.*xx/N);
x1_fit=sig.*Vrs_fit;
y1_fit=sig.*Vrc_fit;
X1_fit= filter(Hd_fit,x1_fit);
Y1_fit=filter(Hd_fit,y1_fit);
S1_fit=2*sqrt(X1_fit.^2+Y1_fit.^2);
 %plot(S1,'r');
 norm_2f_fit=S2_fit./S1_fit;

figure()
plot(t((length(t)/5):end),norm_2f((length(t)/5):end),'color','red');hold on;
plot(t((length(t)/5):end),norm_2f_fit((length(t)/5):end),'color','blue');%fitted WMS


noob=vv_fit./a_fit;
